File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode prev = head;
        int counter = getSize(head) - n;
        
        // if we need to remove the head, just return head's next.
        if(counter==0)
            return head.next;
        
        // loop until we get the node prior to the node we want to remove
        for(int i=0; i<counter-1; i++) {
            prev = prev.next;
        }
        
        // remove that node
        prev.next = prev.next.next;
        
        return head;
    }
    
    private int getSize(ListNode head) {
        int size = 0;
        while(head!=null) {
            size++;
            head = head.next;
        }
        return size;
    }
}

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        tail = head
        
        # get to position where nodeToDeletePrior is n nodes away from end
        while n > 1:
            tail = tail.next
            n-=1
        
        # if the node is the head we want to remove
        if tail.next == None:
            return head.next
            
        # init the counter for the node to remove, but we want the node that is before
        nodeToDeletePrior = head
        tail = tail.next
        
        # Get to the node that you want to delete, but the one before that
        while tail.next != None:
            nodeToDeletePrior = nodeToDeletePrior.next
            tail = tail.next
            
        # Delete the node
        nodeToDeletePrior.next = nodeToDeletePrior.next.next
        
        return head
    
